lms/views/decorators/h_api.py

```diff
@@ -122,3 +124,124 @@ def wrapper(request, jwt):  # pylint: disable=too-many-branches
+            return wrapped(request, jwt)
+
+        try:
+            tool_consumer_instance_guid = request.params["tool_consumer_instance_guid"]
+        except KeyError:
+            raise HTTPBadRequest('Required parameter "tool_consumer_instance_guid" missing from LTI params')
+
+        try:
+            context_id = request.params["context_id"]
+        except KeyError:
+            raise HTTPBadRequest('Required parameter "context_id" missing from LTI params')
+
+        # Show the user an error page if the group hasn't been created yet and
+        # the user isn't allowed to create the group because they aren't an
+        # administrator, instructor or teaching assistant.
+        try:
+            roles = request.params["roles"].lower()
+        except KeyError:
+            raise HTTPBadRequest('Required parameter "roles" missing from LTI params')
+        if not any(role in roles for role in ("administrator", "instructor", "teachingassisstant")):
+            raise HTTPBadRequest('Teacher must configure assignment first')
+
+        group = models.Group.get(request.db, tool_consumer_instance_guid, context_id)
+
+        if group:
+            # The group has already been created in h, nothing more to do here.
+            return wrapped(request, jwt)
+
+        group = models.Group(
+            tool_consumer_instance_guid=tool_consumer_instance_guid,
+            context_id=context_id,
+        )
+
+        # The URL of h's create group API endpoint.
+        create_group_api_url = request.registry.settings["h_api_url"] + "/groups"
+
+        # Our OAuth 2.0 client_id and client_secret for authenticating to the h API.
+        client_id = request.registry.settings["h_client_id"]
+        client_secret = request.registry.settings["h_client_secret"]
+
+        # Generate the name for the new group.
+        try:
+            name = util.generate_group_name(request.params)
+        except MissingContextTitleError:
+            raise HTTPBadRequest('Required parameter "context_title" missing from LTI params')
+
+        authority = request.registry.settings["h_authority"]
+        username = util.generate_username(request.params)
+
+        # Create the group in h.
+        try:
+            response = requests.post(
+                create_group_api_url,
+                auth=(client_id, client_secret),
+                data=json.dumps({"name": name}),
+                headers={
+                    "X-Forwarded-User": "acct:{username}@{authority}".format(username=username, authority=authority),
+                },
+                timeout=1,
+            )
+        except requests.ReadTimeout:
+            raise HTTPGatewayTimeout(explanation="Connecting to Hypothesis failed")
+
+        # Show the user an error page if creating the group in h failed.
+        try:
+            response.raise_for_status()
+        except requests.HTTPError:
+            raise HTTPGatewayTimeout(explanation="Connecting to Hypothesis failed")
```

Comment:
Can you help me understand why `HTTPGatewayTimeout` needs to be raised in two places (and with a duplicated `explanation`)?

Reply:
Fixed by putting both `requests.post()` and `response.raise_for_status()` within the same `try ... except`.